render/gles2/surface.c

```diff
@@ -34,6 +34,35 @@ static bool gles2_surface_attach_pixels(struct wlr_surface_state *surface,
 	return true;
 }
 
+static bool gles2_surface_attach_shm(struct wlr_surface_state *surface,
+		uint32_t format, struct wl_shm_buffer *buffer) {
+	const struct pixel_format *fmt = gl_format_for_wl_format(format);
+	if (!fmt || !fmt->gl_format) {
+		wlr_log(L_ERROR, "No supported pixel format for this surface");
+		return false;
+	}
+	wl_shm_buffer_begin_access(buffer);
+	uint8_t *pixels = wl_shm_buffer_get_data(buffer);
+	int width = wl_shm_buffer_get_width(buffer);
+	int height = wl_shm_buffer_get_height(buffer);
+	int stride = wl_shm_buffer_get_stride(buffer);
+	surface->wlr_surface->width = width;
+	surface->wlr_surface->height = height;
+	surface->wlr_surface->format = format;
+	surface->pixel_format = fmt;
+
+	GL_CALL(glActiveTexture(GL_TEXTURE0));
```

Comment:
I'm not exactly sure why I need this call to glActiveTexture, but it seems to be necessary. Here is my analysis.

`GL_TEXTURE0` is the default active texture when the compositor starts. Upon the first render, the surface is bound with `gles2_surface_bind()` which changes the active texture with `GL_CALL(glActiveTexture(GL_TEXTURE0 + 1));`.

So the first surface is attached with active texture 0 and then it is rendered. Then the second texture is attached with active texture 1 which doesn't seem to work and I just get a black window. In fact, if I attach the first surface with texture 1, I get a black window as well so it seems we must do this with texture 0 active.

Anyway, this works but maybe there is a better way to manage this state. I guess I need to learn more about opengl.